File: Python/String/20.有效的括号.py
# Solution 1
# 大神的解法，类似消消乐
class Solution:
    def isValid(self, s: str) -> bool:
        while '{}' in s or '[]' in s or '()' in s:
            s = s.replace('{}', '')
            s = s.replace('[]', '')
            s = s.replace('()', '')
        return s == ''

# Solution 2
# 用栈
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        left = {'{':'}', '[':']', '(':')'}
        for c in s:
            if c in left:
                stack.append(left[c])
            elif (len(stack) == 0 or c != stack.pop()):
                return False

        return True if len(stack) == 0 else False 

File: Go/String/20-有效的括号.go
package String

// stack
func isValid(s string) bool {
	stack := []byte{}
	for i:=0; i<len(s); i++ {
		if s[i] == '(' || s[i] == '[' || s[i] == '{' {
			stack = append(stack, s[i])
		} else {
			if len(stack) == 0 {
				return false
			}
			switch s[i] {
			case ')':
				if stack[len(stack)-1] != '(' {
					return false
				}
			case ']':
				if stack[len(stack)-1] != '[' {
					return false
				}
			case '}':
				if stack[len(stack)-1] != '{' {
					return false
				}
			}
			stack = stack[:len(stack)-1]
		}
	}
	if len(stack) != 0 {
		return false
	}
	return true
}
